enhancements/subscription-content/subscription-content-access.md

```diff
@@ -0,0 +1,242 @@
+tells yum/dnf where to access content from.
+
+Yum repository definitions are specific to the major version of RHEL that the running container is
+based on - a RHEL 8-based container is not compatible with RHEL 7 content.
+
+This enhancement proposal will detail at a high level the functions of:
+
+A cluster-wide subscription manager to generate subscription content access data A projected
+resource CSI driver to share resources across OpenShift namespaces A subscription content webhook to
+generate the necessary volume mounts to add subscription content access data to containers
+Extensions to propagate subscription content access data into OpenShift builds.
+
+#### Cluster Subscription Manager
+
+The cluster subscription manager will take as input the `cloud.redhat.com` pull secret, which
+already provides authentication for the cluster and provides the means to register clusters with Red
+Hat’s cloud manager. The pull secret is then used to obtain the following: The `redhat.repo` yum
+repo configurations for all supported major versions of RHEL. Today these will be RHEL 7 and 8 -
+later major versions of RHEL must be anticipated. The Simple Content Access certificate for the
+customer, which will serve as the entitlement key.
+
+As RHSM does today on RHEL hosts, the cluster subscription manager will be responsible for
+generating `redhat.repo` definitions for the currently supported RHEL major versions and saving it
+on the cluster. Each `redhat.repo` file will be saved in a ConfigMap within the `openshift-config`
+namespace, named `redhat-repo-rhel${n}`, where `${n}` is the major version of RHEL associated with
+the `redhat.repo` file.
+
+The cluster subscription manager will also need to obtain the entitlement keys associated with the
+attached OpenShift subscription. It is assumed that OpenShift entitlements grant access to RHEL
+content, but may not grant access to layered content such as EAP or Application Runtimes. This
+content must be saved in a Secret within the `openshift-config` namespace, named
+`etc-pki-entitlement`.
+
+The full details of this component will be specified in a subsequent enhancement proposal.
+
+#### Projected-resource CSI Driver
+
+Once the entitlement key(s) and yum repository definitions are stored on the cluster, the next step
+is to share these definitions across namespaces. This task will fall to the projected-resource
+containers storage interface (CSI) driver. This will be a special storage driver that allows
+ConfigMaps and Secrets to be shared across namespaces via standard pod volume mounts. Access to
+shared Secrets and ConfigMaps will be controlled via namespace-scoped Custom Resources. The
+projected-resource CSI driver will be responsible for conducting a subject access review (SAR) check
+to ensure that the service account for the pod has permission to mount any projected resources.
+
+The full details of this component are specified in the
```

Comment:
Noting here so I don’t forget later, but is the csi driver itself only able to read the one named secret?

Reply:
If it helps @derekwaynecarr the sample yaml we settled on based on prompting from @deads2k : 

```
kind: ProjectedResource
apiVersion: projected-resource.storage.openshift.io/v1
metadata:
  name: shared-cool-secret
spec:
  backingResource:
    kind: Secret
    apiVersion: v1
    namespace: ns-one
    name: cool-secret
  description: "used to access my internal git repo, RHEL entitlements, or something else cool"
status:
  conditions:
  ...
```

Reply:
Furthermore we will need a ConfigMap to mount in the right `redhat.repo` file (these need to be different if you're running a rhel7 or rhel8 container).

@gabemontero food for thought - k8s supports a `ProjectedVolume` which lets you combine `Secrets` and `ConfigMaps` [1]. We should separately discuss if this is something we want to support in the CSI driver API.

[1] https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#volumeprojection-v1-core

Reply:
interesting @adambkaplan ... yeah add it as a note  to the appropriate Jira in this space you recently opened